core/src/main/java/org/infinispan/remoting/responses/SuccessfulResponse.java

```diff
@@ -1,85 +1,99 @@
 package org.infinispan.remoting.responses;
 
-import java.io.IOException;
-import java.io.ObjectInput;
-import java.io.ObjectOutput;
-import java.util.Set;
-
-import org.infinispan.commons.marshall.AbstractExternalizer;
-import org.infinispan.commons.util.Util;
-import org.infinispan.marshall.core.Ids;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+
+import org.infinispan.commons.marshall.ProtoStreamTypeIds;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.marshall.protostream.impl.MarshallableArray;
+import org.infinispan.marshall.protostream.impl.MarshallableCollection;
+import org.infinispan.marshall.protostream.impl.MarshallableMap;
+import org.infinispan.marshall.protostream.impl.MarshallableObject;
+import org.infinispan.protostream.annotations.ProtoFactory;
+import org.infinispan.protostream.annotations.ProtoTypeId;
 
 /**
  * A successful response
  *
  * @author Manik Surtani
  * @since 4.0
  */
+@ProtoTypeId(ProtoStreamTypeIds.SUCCESSFUL_RESPONSE)
 public class SuccessfulResponse<T> extends ValidResponse {
```

Comment:
Since it is in the critical path of all operations, I would convert this to an interface and have different implementation for each response value:

* empty/null
* object
* array
* bytes
* collection
* map

Reply:
Same idea for UnsuccessfulResponse.

Reply:
Good idea. This actually allows me to revert some of my earlier changes to the interceptor. I've pushed a commit with the change.